File: pages/20260819100800_grip_algebra/analysis/julia/combinatorics.jl
function fubini_number(object_count::Integer)
    object_count >= 0 || throw(ArgumentError("object count must be nonnegative"))
    values = [1]
    for size in 1:object_count
        push!(
            values,
            sum(binomial(size, block_size) * values[size - block_size + 1] for block_size in 1:size),
        )
    end
    return values[end]
end

function companion_release_count(packet_sizes)
    sizes = collect(Int, packet_sizes)
    all(size >= 0 for size in sizes) || throw(
        ArgumentError("packet sizes must be nonnegative"),
    )
    total = sum(sizes)
    iszero(total) && return 0 // 1
    return sum(size * (size - 1) for size in sizes) // total
end

function interior_state_count(object_count::Integer, buffer_events::Integer)
    object_count >= 0 || throw(ArgumentError("object count must be nonnegative"))
    buffer_events >= 0 || throw(ArgumentError("buffer events must be nonnegative"))
    return sum(
        binomial(object_count, held)
        for held in 0:object_count
        if min(held, object_count - held) > buffer_events;
        init=0,
    )
end

function mixed_cycle_rank(object_count::Integer)
    object_count >= 3 || throw(ArgumentError("connected mixed graph starts at three objects"))
    return (object_count - 2) * 2^(object_count - 1) - 2 * object_count + 3
end

function mixed_cycle_rank_enumerated(object_count::Integer)
    object_count >= 3 || throw(ArgumentError("connected mixed graph starts at three objects"))
    mixed = 1:(2^object_count - 2)
    allowed = Set(mixed)
    edges = 0
    for state in mixed
        for bit in 0:(object_count - 1)
            neighbor = state ⊻ (1 << bit)
            if neighbor in allowed && neighbor > state
                edges += 1
            end
        end
    end
    return edges - length(mixed) + 1
end

File: pages/20260819100800_grip_algebra/analysis/ruby/combinatorics.rb
module GripAnalysis
  module Combinatorics
    module_function

    def fubini_number(object_count)
      unless object_count.is_a?(Integer) && object_count >= 0
        raise ArgumentError, "object count must be a nonnegative integer"
      end

      values = [1]
      (1..object_count).each do |size|
        values << (1..size).sum { |block_size| binomial(size, block_size) * values.fetch(size - block_size) }
      end
      values.fetch(object_count)
    end

    def companion_release_count(packet_sizes)
      sizes = Array(packet_sizes)
      unless sizes.all? { |size| size.is_a?(Integer) && size >= 0 }
        raise ArgumentError, "packet sizes must be nonnegative integers"
      end

      total = sizes.sum
      return 0.to_r if total.zero?

      Rational(sizes.sum { |size| size * (size - 1) }, total)
    end

    def interior_state_count(object_count, buffer_events)
      unless object_count.is_a?(Integer) && object_count >= 0
        raise ArgumentError, "object count must be a nonnegative integer"
      end
      unless buffer_events.is_a?(Integer) && buffer_events >= 0
        raise ArgumentError, "buffer events must be a nonnegative integer"
      end

      (0..object_count).sum do |held|
        next 0 if [held, object_count - held].min <= buffer_events

        binomial(object_count, held)
      end
    end

    def mixed_cycle_rank(object_count)
      unless object_count.is_a?(Integer) && object_count >= 3
        raise ArgumentError, "connected mixed graph starts at three objects"
      end

      (object_count - 2) * 2**(object_count - 1) - 2 * object_count + 3
    end

    def binomial(total, selected)
      return 0 if selected.negative? || selected > total

      selection = [selected, total - selected].min
      (1..selection).reduce(1) do |product, index|
        product * (total - selection + index) / index
      end
    end
  end
end
